### PR TITLE
Loki: Add parsing of query to visual query

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
This PR adds parsing between text query and visual query. I

What is missing and could be part of the next PR: 
- [x] unwrap
- [x]  grouping (e.g sum_by), but this is currently not supported in visual query editor either
- [x]  implement line_filter in visual query editor
- [x]  implement label_filter in visual query editor

Part of: 

https://github.com/grafana/grafana/issues/44253

### CODE REVIEW

**Comment by torkelo** [comment]
If I use ticks for string they get added as ticks inside strings in the model 
Screenshot from 2022-03-22 09-04-58
Screenshot from 2022-03-22 09-36-44

**Comment by ivanahuckova** [comment]
> If I use ticks for string they get added as ticks inside strings in the model

Fixed. I have also added functionality that unescapes `\` if user uses `"` and is escaping it.

File: public/app/plugins/datasource/loki/querybuilder/LokiQueryModeller.ts [hunk 1]
```diff
@@ -27,8 +27,13 @@ export class LokiQueryModeller extends LokiAndPromQueryModellerBase<LokiVisualQu
...
   renderQuery(query: LokiVisualQuery) {
     let queryString = `${this.renderLabels(query.labels)}`;
-    queryString = this.renderOperations(queryString, query.operations);
-    queryString = this.renderBinaryQueries(queryString, query.binaryQueries);
+    try {
+      queryString = this.renderOperations(queryString, query.operations);
+      queryString = this.renderBinaryQueries(queryString, query.binaryQueries);
+    } catch (err) {
+      // TODO: For now we just log it
+      console.error(err);
```
[the remark this answers is not in the document]
Comment (aocenas) [reply public/app/plugins/datasource/loki/querybuilder/LokiQueryModeller.ts:35]: oh ok but this is a different part. I thought we talked about the parsing. Here in rendering erroring out seems a bit weird. like you have an operation in the vis query object but at the same time, you don't know how to render it? In that case, how did it even get into that object? If it was added by the user I would assume we have a definition for it (and so also render code) or if it was parsed from string but we don't have a definition that could be seen as a parsing bug because we also don't know what UI to use for such operation.

File: public/app/plugins/datasource/loki/querybuilder/components/LokiQueryEditorSelector.test.tsx [hunk 2]
```diff
@@ -125,29 +132,30 @@ describe('LokiQueryEditorSelector', () => {
...
-  //     />
-  //   );
-
-  //   await screen.findByText('test_metric');
-  //   expect(screen.getByText('host.docker.internal:3000')).toBeInTheDocument();
-  //   expect(screen.getByText('Rate')).toBeInTheDocument();
-  //   expect(screen.getByText('$__interval')).toBeInTheDocument();
-  // });
+  it('parses query when changing to builder mode', async () => {
+    jest.spyOn(console, 'error').mockImplementation();
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/querybuilder/components/LokiQueryEditorSelector.test.tsx:136]: Why is this here?

File: public/app/plugins/datasource/loki/querybuilder/components/LokiQueryEditorSelector.tsx [hunk 3]
```diff
@@ -18,10 +19,24 @@ export const LokiQueryEditorSelector = React.memo<LokiQueryEditorProps>((props)
...
   const [visualQuery, setVisualQuery] = useState<LokiVisualQuery>(query.visualQuery ?? getDefaultEmptyQuery());
+  const [parseModalOpen, setParseModalOpen] = useState(false);
+  const [pendingChange, setPendingChange] = useState<LokiQuery | undefined>(undefined);
 
   const onEditorModeChange = useCallback(
     (newMetricEditorMode: QueryEditorMode) => {
-      onChange({ ...query, editorMode: newMetricEditorMode });
+      const change = { ...query, editorMode: newMetricEditorMode };
+      if (newMetricEditorMode === QueryEditorMode.Builder) {
+        const result = buildVisualQueryFromString(query.expr);
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/querybuilder/components/LokiQueryEditorSelector.tsx:30]: For some other PR probably, but I think this (parsing + error handling modal) could be moved into a separate hook and reused.

File: public/app/plugins/datasource/loki/querybuilder/components/LokiQueryEditorSelector.test.tsx (hunk 2 above)
Comment (aocenas) [reply public/app/plugins/datasource/loki/querybuilder/components/LokiQueryEditorSelector.test.tsx]: Seems like there was unchecked async setState which triggered an error, so fixed that and this is not needed anymore

**Review by torkelo — APPROVED** [review]
